### PR TITLE
calendar: implement alaram widget to poll from bus.

### PR DESCRIPTION
**Pull request by apa-odoo**
Task: https://www.odoo.com/web#id=19576&view_type=form&model=project.task&action=327

Pad: https://pad.odoo.com/p/r.vLPf7RQxElNE4MlN

### CODE REVIEW

**Comment by antonylesuisse** [comment]
Dont touch bus. When there is a change to could affect the next_event_time. Broadcast an a message on the bus to the concerned users. In js when this event is receive do a rpc to update the next_event_time (the rpc we did every 5min).

File: addons/calendar/static/src/js/base_calendar.js [hunk 1]
```diff
@@ -179,60 +182,51 @@ var CalendarNotification = Notification.extend({
...
-            if(err.code === -32098) {
-                // Prevent the CrashManager to display an error
-                // in case of an xhr error not due to a server error
-                ev.preventDefault();
-            }
-        });
+        this._super(parent);
+        this.bus = bus.bus;
+        this.bus.last = 0;
+        this.bus.on("notification", this, this.on_notification);
```
Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: Is there a need to store it in this ?

File: addons/calendar/static/src/js/base_calendar.js [hunk 2]
```diff
@@ -8,8 +8,11 @@ var Dialog = require('web.Dialog');
 var form_common = require('web.form_common');
 var Model = require('web.DataModel');
 var Notification = require('web.notification').Notification;
+var NotificationManager = require('web.notification').NotificationManager;
```
Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: Is there any use of NotificationManager ?

File: addons/calendar/static/src/js/base_calendar.js [hunk 3]
```diff
@@ -187,52 +190,44 @@ var CalendarNotification = Notification.extend({
...
+        var channel = notification[0];
+        var message = notification[1];
+
+        if (message.event_id){
+            setTimeout(function() {
+                self.getParent().notification_manager.display(new CalendarNotification(self.getParent().notification_manager, message));
+            }, message.timer * 1000);
+        }
+    }
+});
```
Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: Where you differentiate whether the  notification is for your channel(calendar.alarm) or not, we do not want to display all notification here, may be you should check channel[1] === 'calendar.alarm'

File: addons/calendar/static/src/js/base_calendar.js [hunk 4]
```diff
@@ -187,52 +190,44 @@ var CalendarNotification = Notification.extend({
...
+        this.bus.last = 0;
+        this.bus.on("notification", this, this.on_notification);
     },
-    check_notifications: function() {
-        var self = this;
-        this.get_next_notif();
-        this.intervalNotif = setInterval(function() {
-            self.get_next_notif();
-        }, 5 * 60 * 1000);
+    get_notif_box: function(me) {
```
Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: Is this method used anywhere ?

File: addons/calendar/static/src/js/base_calendar.js [hunk 5]
```diff
@@ -187,52 +190,44 @@ var CalendarNotification = Notification.extend({
...
     },
+    on_notification: function(notification) {
+        var self = this;
+        var channel = notification[0];
+        var message = notification[1];
+
+        if (message.event_id){
+            setTimeout(function() {
+                self.getParent().notification_manager.display(new CalendarNotification(self.getParent().notification_manager, message));
+            }, message.timer * 1000);
```
Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: May be better to add condition to not show same notification twice:
```if(self.$(".eid_" + res.event_id).length === 0)```

File: addons/calendar/static/src/js/base_calendar.js [hunk 6]
```diff
@@ -187,52 +190,44 @@ var CalendarNotification = Notification.extend({
...
-        .fail(function(err, ev) {
-            if(err.code === -32098) {
-                // Prevent the CrashManager to display an error
-                // in case of an xhr error not due to a server error
-                ev.preventDefault();
-            }
-        });
+        this._super(parent);
+        this.bus = bus.bus;
+        this.bus.last = 0;
```
Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: Is there any specific reason to set last = 0, note that, if last is 0 then first poll will return messages of last 50 seconds(by default)

File: addons/calendar/static/src/js/base_calendar.js (hunk 2 above)
Comment (apa-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: No, I remove it. by mistake it's remain. thanks !!

File: addons/calendar/static/src/js/base_calendar.js (hunk 1 above)
Comment (apa-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: No, it not required, but in old prototype bus use this way so. I remove it.

File: addons/calendar/static/src/js/base_calendar.js (hunk 3 above)
Comment (apa-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: if(message.event_id) is for differentiate calendar event, but channel is also use for it.

File: addons/calendar/static/src/js/base_calendar.js (hunk 4 above)
Comment (apa-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: No, I remove it. Thanks !!

File: addons/calendar/static/src/js/base_calendar.js (hunk 3 above)
Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: I am not able to find where you differentiate whether notification is from calendar.alarm channel or not

File: addons/calendar/static/src/js/base_calendar.js (hunk 5 above)
Comment (apa-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: <code>self.$(".eid_" + res.event_id).length</code> always return 0. 
<code>$(".eid_" + res.event_id).length</code> i use it for prevent notification twice on same event.

Comment (msh-odoo) [inline addons/calendar/static/src/js/base_calendar.js]: That is because now you created a CalendarAlarm widget so you will not get it in current element, may be you need to do self.getParent().$(...), but do not use global selector

File: addons/calendar/calendar.py [hunk 7]
```diff
@@ -425,18 +416,18 @@ def do_check_alarm_for_one_date(self, cr, uid, one_date, event, event_maxdelta,
...
-        now = openerp.fields.Datetime.to_string(datetime.now())
+    def get_next_notification(self, cr, uid, context=None):
+        all_notif = []
 
         icp = self.pool['ir.config_parameter']
-        last_notif_mail = icp.get_param(cr, SUPERUSER_ID, 'calendar.last_notif_mail', default=False) or now
 
         try:
-            cron = self.pool['ir.model.data'].get_object(cr, SUPERUSER_ID, 'calendar', 'ir_cron_scheduler_alarm', context=context)
+            cron = self.pool['ir.model.data'].get_object(cr, uid, 'calendar', 'ir_cron_scheduler_alarm', context=context)
```
Comment (msh-odoo) [inline addons/calendar/calendar.py]: Why this change ?

File: addons/calendar/calendar.py [hunk 8]
```diff
@@ -455,62 +446,44 @@ def get_next_mail(self, cr, uid, context=None):
...
+                    if alert['alarm_type'] == 'email':
+                        self.do_mail_reminder(cr, uid, alert, context=context)
+                    if alert['alarm_type'] == 'notification':
                         all_notif.append(self.do_notif_reminder(cr, uid, alert, context=context))
-        return all_notif
+        if all_notif:
+            self.pool['bus.bus'].sendmany(cr, uid, all_notif)
+        else:
+            now = openerp.fields.Datetime.to_string(datetime.now())
+            icp.set_param(cr, SUPERUSER_ID, 'calendar.last_notif_mail', now)
```
Comment (msh-odoo) [inline addons/calendar/calendar.py]: It means if there are some notifications then do not set last_notif_mail param, why not always set last_notif_mail param ?

**Comment by hmo-odoo** [comment]
this feature implemented in different way. see this commit: https://github.com/odoo/odoo/commit/488e9f299deac6c0b0c7adc6fd68553ffa56e579
